### PR TITLE
Parameterized user-defined attributes

### PR DESCRIPTION
**Pull request by Kha**
Add a parameter parser to `user_attribute`, store the parsed result in reflected form and provide a typed API to retrieve it. Implement a `[derive]` attribute on top of that.

/cc @rlewis1988, https://github.com/formalabstracts/formalabstracts/pull/49

### CODE REVIEW

File: library/init/meta/derive.lean [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+| (h::hs) :=
+do success ← h p n,
+   when (¬success) $
+     try_handlers hs
+
+@[user_attribute] meta def derive_attr : user_attribute unit (list pexpr) :=
+{ name := `derive, descr := "automatically derive typeclass instances",
+  parser := pexpr_list_or_texpr,
+  after_set := some (λ n _ _,
+  do ps ← derive_attr.get_param n,
```
Comment (Kha) [inline library/init/meta/derive.lean:39]: Note the recursive use of `derive_attr` -- unbounded recursion can be helpful indeed.

**Comment by gebner** [comment]
Nice!  Is there a way to set a parameterized user attribute from Lean?

**Comment by gebner** [comment]
Hmm, the following would fail, right?
```lean
inductive foo | bar | baz
attribute [derive decidable_eq] foo
attribute [derive inhabited] foo
attribute [derive has_reflect] foo
```

**Comment by Kha** [comment]
@gebner 
> Is there a way to set a parameterized user attribute from Lean?

Oh, I guess there should be! I was actually wondering whether we should change the system attributes to persist a single `expr` as well so that we can finally set them from Lean.

**Comment by Kha** [comment]
@gebner That will actually work (apart from the missing derive handler for `has_reflect`), but only the last `[derive]` instance will be persisted (which I guess doesn't matter for that attribute).

**Comment by Kha** [comment]
That is,

```lean
inductive foo | bar | baz
attribute [derive decidable_eq] foo
attribute [derive inhabited] foo

#check foo.decidable_eq
│ foo.decidable_eq : decidable_eq foo

#check foo.inhabited
│ foo.inhabited : inhabited foo

#print foo
│ @[derive list.cons.{0} pexpr ``(inhabited) (list.nil.{0} pexpr)]
│ inductive foo : Type                                            
│ constructors:                                                   
│ foo.bar : foo                                                   
│ foo.baz : foo                                                   
```

(We should probably use an `io_state_stream` for attribute printing...)

**Comment by leodemoura** [comment]
@kha please do not merge PRs which are not small bug fixes.
I wanted to revise this PR before merging it.

**Pull request merged**